src/Driver.cpp

```diff
@@ -535,3 +555,86 @@ void Driver::GTES(void)
     waitResponseOK(command);
     modeMsgManager(command);
 }
+
+// Switch to COMMAND mode.
+void Driver::switch2CMDmode(void)
+{
+    string command = "ATC";
+    sendCommand(command);
+    waitResponseOK(command);
+    modeMsgManager(command);
+}
+
+// Switch to DATA mode.
+void Driver::switch2DATAmode(void)
+{
+    string command = "AT0";
+    sendCommand(command);
+    modeMsgManager(command);
+}
+
+// Reset device, drop data and/or instant message.
+void Driver::resetDevice(ResetType const &type)
+{
+    stringstream command;
+    command << "ATZ" << type;
+    sendCommand(command.str());
+    waitResponseOK(command.str());
+}
+
+// Pop out RawData from queueRawData.
+std::string Driver::getRawData(void)
+{
+    if(!queueRawData.empty())
+    {
+        std::string ret = queueRawData.front();
+        queueRawData.pop();
+        return ret;
+    }
+    throw runtime_error("Driver.cpp getRawData: queueRawData is empty.");
+}
+
+// verify if queueRawData has raw data.
+bool Driver::hasRawData(void)
+{
+    return !queueRawData.empty();
+}
+
+// Pop out Notification from queueNotification.
+NotificationInfo Driver::getNotification(void)
+{
+    if(!queueNotification.empty())
+    {
+        NotificationInfo ret = queueNotification.front();
+        queueNotification.pop();
+        return ret;
+    }
+    throw runtime_error("Driver.cpp getNotification: queuenNotification is empty.");
+}
+
+// verify if queueNotification has any notification.
+bool Driver::hasNotification(void)
+{
+    return !queueNotification.empty();
+}
+
+// Get mode of operation.
+OperationMode Driver::getMode(void)
+{
+    return mode;
+}
+
+base::Quaterniond Driver::eulerToQuaternion(const base::Vector3d &eulerAngles)
```

Comment:
Ohhhh. Yes. There is much better.

base/types has bryan angles-to-quaternion, but they might not be the exact same convention than what is used by the USBL. However, you should definitely get your inspiration from there. Basically, euler-to-quaternion is a composition of rotations around axis, so we express it exactly that way in Eigen as well:

~~~
q = Eigen::AngleAxisd(rot0, Eigen::Vector3d::UnitZ()) *
      Eigen::AngleAxisd(rot1, Eigen::Vector3d::UnitY()) *
      Eigen::AngleAxisd(rot2, Eigen::Vector3d::UnitX());
~~~

Note, the exact convertion depends on the convention used by the USBL ...